BraveRewards/Ledger/Data/BATLedgerDatabase.mm

```diff
@@ -0,0 +1,555 @@
+    auto mi = [[MediaPublisherInfo alloc] initWithEntity:MediaPublisherInfo.entity
+                          insertIntoManagedObjectContext:context];
+    updateInfo(mi);
+  }];
+}
+
+#pragma mark - Recurring Tips
+
++ (NSArray<BATPublisherInfo *> *)recurringTipsForMonth:(BATActivityMonth)month year:(int)year
+{
+  const auto context = DataController.viewContext;
+  const auto fetchRequest = RecurringDonation.fetchRequest;
+  fetchRequest.entity = [NSEntityDescription entityForName:NSStringFromClass(RecurringDonation.class)
+                                    inManagedObjectContext:context];
+  fetchRequest.predicate = [NSPredicate predicateWithFormat:@"month = %d AND year = %d AND category = %d",
+                            month, year, BATRewardsCategoryOneTimeTip];
+  
+  const auto frc = [[NSFetchedResultsController alloc] initWithFetchRequest:fetchRequest
+                                                       managedObjectContext:context
+                                                         sectionNameKeyPath:nil
+                                                                  cacheName:nil];
+  
+  NSError *error;
+  if (![frc performFetch:&error]) {
+    NSLog(@"%@", error);
+  }
+  
+  const auto publishers = [[NSMutableArray<BATPublisherInfo *> alloc] init];
+  for (RecurringDonation *rd in frc.fetchedObjects) {
+    auto info = [[BATPublisherInfo alloc] init];
+    info.id = rd.publisherID;
+    info.name = rd.publisher.name;
+    info.url = rd.publisher.url.absoluteString;
+    info.faviconUrl = rd.publisher.faviconURL.absoluteString;
+    info.weight = rd.amount;
+    info.reconcileStamp = rd.addedDate;
+    info.verified = rd.publisher.verified;
+    info.provider = rd.publisher.provider;
+    [publishers addObject:info];
+  }
+  return publishers;
+}
+
++ (void)insertOrUpdateRecurringTipWithPublisherID:(NSString *)publisherID
+                                           amount:(double)amount
+                                        dateAdded:(uint32_t)dateAdded
+{
+  if (publisherID.length == 0) {
+    return;
+  }
+  
+  void (^updateInfo)(RecurringDonation *) = ^(RecurringDonation *rd) {
+    rd.publisherID = publisherID;
+    rd.amount = amount;
+    rd.addedDate = dateAdded;
+    rd.publisher = [self getPublisherInfoWithID:publisherID];
+  };
+  
+  const auto rd = [self getRecurringDonationWithPublisherID:publisherID];
+  if (rd) {
+    updateInfo(rd);
+    return;
+  }
+  
+  [DataController.shared performOnContext:nil task:^(NSManagedObjectContext * _Nonnull context) {
+    auto rd = [[RecurringDonation alloc] initWithEntity:RecurringDonation.entity
+                         insertIntoManagedObjectContext:context];
+    updateInfo(rd);
+  }];
+}
+
++ (BOOL)removeRecurringTipWithPublisherID:(NSString *)publisherID
+{
+  const auto rd = [self getRecurringDonationWithPublisherID:publisherID];
+  if (!rd) {
+    return NO;
+  }
+  [DataController.viewContext deleteObject:rd];
```

Comment:
use `performWithContext`, we should avoid deletes on main thread

this can also be tricky since we expect to return a `bool`